src/components/authorize/util.ts

```diff
@@ -0,0 +1,211 @@
+      'count',
+      'create',
+      'find',
+      'findOne',
+      'findById',
+      'replaceById',
+      'updateById',
+      'deleteById',
+      'updateAll',
+    ];
+
+    // Controller not extended from any class
+    if (methodsParentsMethods.includes('__proto__')) {
+      return this.generatePermissions({
+        methods: methodsChildClass,
+        permissionSubject: parentPermission.subject,
+        parentId: parentPermission.id,
+        allPermissionDecoratorData,
+      });
+    }
+
+    // Controller is extended from CrudController
+    return this.generatePermissions({
+      methods: union(defaultPermissions, methodsChildClass),
+      permissionSubject: parentPermission.subject,
+      parentId: parentPermission.id,
+      allPermissionDecoratorData,
+    });
+  };
+
+  generatePermissionRecords(opts: {
+    controller: Function;
+    parentPermission: Permission;
+    permissionRepository: PermissionRepository;
+    allPermissionDecoratorData: MetadataMap<unknown> | object;
+  }) {
+    const { controller, parentPermission, allPermissionDecoratorData } = opts;
+    const permissionRecords: IPermission[] = [];
+    const controllerPrototype = controller.prototype;
+    const methodsChildClass = this.getMethodsClass(controllerPrototype);
+
+    const parentClass = Reflect.getPrototypeOf(controllerPrototype) as object;
+    const methodsParentsMethods = this.getMethodsClass(parentClass);
+
+    permissionRecords.push(
+      ...this.generatePermissionBaseInherit({
+        methodsParentsMethods,
+        methodsChildClass,
+        parentPermission,
+        allPermissionDecoratorData,
+      }),
+    );
+
+    return permissionRecords;
+  }
+
+  async updatePermissionByChangeMethodName(
+    permissionSubject: string,
+    allPermissionDecoratorData: Record<string, { idx: string }>,
+    permissionRepository: PermissionRepository,
+  ) {
+    if (!Object.values(allPermissionDecoratorData).length) {
+      return;
+    }
+
+    const allPermissionDecorators = Object.entries(allPermissionDecoratorData);
+
+    for (const [key, value] of allPermissionDecorators) {
+      const permissionsFound: Permission[] = await permissionRepository.find({
+        where: {
+          subject: permissionSubject,
+          code: {
+            neq: `${permissionSubject}.*`,
+          },
+        },
+      });
+
+      for (const p of permissionsFound) {
+        if (!p?.details?.idx || p?.details?.idx !== value.idx) {
+          continue;
+        }
+
+        await permissionRepository.updateById(p.id, {
+          ...p,
+          code: `${permissionSubject}.${key}`,
+          details: { ...value },
+        });
+      }
+    }
+  }
+
+  async startMigration(opts: { permissionRepository: any; controllers: object }) {
+    const { permissionRepository, controllers } = opts;
+    const permissions: IPermission[] = [];
+
+    for (const controller of Object.values(controllers)) {
+      const controllerClass = controller as Function;
```

Comment:
Dont treat a class as a function

Reply:
Actually, I find out that the class here should be type Function. Because, I need to access some property like this:
`      const controllerClass = controller as Function;
      const permissionSubject = controllerClass.name.replace(/Controller/g, '');
      const controllerPrototype = controllerClass.prototype;
`